### PR TITLE
feat(embodied): add Evo-1 flow-matching VLA model

### PR DESCRIPTION
**Pull request by DorayakiLin**
# PR: feat(embodied): add Evo-1 (InternVL3 + flow-matching) VLA model support

> Target repo: `RLinf/RLinf`. Paste the body below into the GitHub PR.

## Summary

This PR adds **Evo-1** (`evo1-flash`) — a compact **InternVL3-1B VLM + flow-matching (DiT)
action head** vision-language-action model — as a first-class embodied model in RLinf, with
**SFT** and **GRPO** support on LIBERO. The integration is **fully additive**, and on
LIBERO-Spatial, GRPO from an Evo-1 SFT checkpoint yields a **clean, monotonic success-rate
improvement**.

## Highlights

- **Compact & practical** — a ~0.77B VLA that trains and runs on a single node.
- **Native, in-process integration** — embedded in RLinf's memory space for zero-latency,
  tensor-level interaction (no server / RPC hop).
- **Clean RL gain** — on a single LIBERO-Spatial task, ``success_at_end`` rises monotonically
  **0.58 → 0.86** over ~40 GRPO steps with stable training dynamics.
- **Minimal & safe to merge** — purely additive; existing code paths and image builds are untouched.

## What's added

**Model** (`rlinf/models/embodiment/evo1/`)
- `evo1_action_model.py` (policy wrapper), `evo1_rl_head.py` (flow-matching SDE-replay
  log-prob), `sft_builder.py`, `utils/`, `__init__.py`.

**Registration & wiring** (additive, evo1-gated)
- `rlinf/config.py` (`SupportedModel.EVO1`), `rlinf/models/__init__.py` (builder),
  rollout + SFT worker branches.

**Configs**
- `examples/embodiment/config/model/evo1.yaml`, `libero_spatial_grpo_evo1.yaml` (GRPO),
  `examples/sft/config/libero_sft_evo1.yaml` + `model/evo1.yaml` (SFT),
  `evaluations/libero/libero_spatial_evo1_eval.yaml` (standalone eval).

**Install & docs**
- `requirements/install.sh` (`install_evo1_model`); `docs/.../examples/embodied/evo1.rst`
  (EN + ZH) + gallery toctree entry.

**Test**
- `tests/e2e_tests/embodied/libero_spatial_grpo_evo1.yaml` + a CI e2e job.

## Recipe (shipped config defaults)

- `adv_type=grpo`, actor-only; **freeze the VLM, train only the flow-matching action head**
  (~122M / 775M ≈ 15.7% of parameters).
- `logprob_type=token_level` (per-dim; matches Evo-1's 14-step executed action chunk).
- `noise_method=flow_sde`, `noise_level=0.5`; `group_size=8`; `update_epoch=2`;
  `clip_ratio=0.2/0.28`.

## Results (LIBERO-Spatial, GRPO from an Evo-1 SFT checkpoint, n=64)

| Setting | SFT | RLinf-GRPO |
|---|---|---|
| Single task (`success_at_end`) | 0.58 | **0.86** |
| LIBERO-Spatial suite (`success_once`, best ckpt) | 0.656 | **0.750** |

The single-task curve is monotonic (0.58 → 0.72 → 0.74 → **0.86**) with stable KL and
gradient norms — a clean demonstration that GRPO fine-tuning improves the Evo-1 policy.

## How to run

```bash
bash requirements/install.sh embodied --model evo1 --env libero --use-mirror
export ROBOT_PLATFORM=LIBERO EVO1_REPO_PATH=/path/to/Evo-1
bash examples/embodiment/run_embodiment.sh libero_spatial_grpo_evo1
```

To reproduce the single-task result quickly, add
``+env.train.task_id_filter=[0] +env.eval.task_id_filter=[0]``.

## Checklist

- [x] Model code + `SupportedModel.EVO1` registration + builder
- [x] Rollout + SFT worker branches
- [x] Example SFT + GRPO configs + standalone eval config
- [x] `requirements/install.sh` (`install_evo1_model`)
- [x] Docs page `evo1.rst` (EN + ZH) + gallery toctree
- [x] e2e test config + CI job
- [x] Commits signed off (DCO)

## Notes for maintainers

- **Purely additive**: every change to an existing file is a new, evo1-gated addition —
  nothing existing changes behavior.
- Docker image baking is intentionally left out to keep this PR minimal; Evo-1 installs
  cleanly via `install.sh --model evo1 --env libero`.
- The e2e job expects an Evo-1 LIBERO checkpoint on the runner; `arm_key` / `dataset_key`
  must match the checkpoint's `norm_stats.json` top-level key(s).

### CODE REVIEW

File: examples/embodiment/config/libero_spatial_grpo_evo1.yaml [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    lr: 5.0e-6
+    value_lr: 1.0e-4
+    adam_beta1: 0.9
+    adam_beta2: 0.95
+    adam_eps: 1.0e-08
+    weight_decay: 0.01
+    clip_grad: 1.0
+
+  fsdp_config:
+    strategy: "fsdp"
```
Comment (thereAreDemonsNearby) [inline examples/embodiment/config/libero_spatial_grpo_evo1.yaml:148]: Should set use_orig_params=True as you said in evo1.yaml

File: examples/embodiment/config/libero_spatial_grpo_evo1.yaml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+rollout:
+  group_name: "RolloutGroup"
+  generation_backend: "huggingface"
+  recompute_logprobs: False      # GRPO uses the rollout-time prev_logprobs
+  unnorm_key: libero_10          # unused by Evo-1 (its Normalizer keys on arm/dataset_key)
+  enable_offload: False
+  pipeline_stage_num: 1
+
+  model:
+    model_path: "/data/oss_bucket_0/data/taolin/checkpoints/Evo1/Evo1_libero"
```
Comment (thereAreDemonsNearby) [inline examples/embodiment/config/libero_spatial_grpo_evo1.yaml]: Shouldn't use machine specific paths. Please find and replace them all.

File: rlinf/models/embodiment/evo1/evo1_action_model.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+            or head_steps
+            or 32
+        )
+
+    # ------------------------------------------------------------------ #
+    # FSDP auto-wrap: transformer layer classes to shard/wrap.
+    # InternVL3-1B embedder = InternViT (InternVisionEncoderLayer) + Qwen2 LLM
+    # (Qwen2DecoderLayer); the flow-matching action head uses BasicTransformerBlock.
+    # ------------------------------------------------------------------ #
+    _no_split_modules = [
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/evo1/evo1_action_model.py]: _no_split_modules is defined twice in this file

File: examples/embodiment/config/libero_spatial_grpo_evo1.yaml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  huber_delta: 10.0
+
+  gamma: 0.99
+  gae_lambda: 0.95
+
+  filter_rewards: True
+  rewards_lower_bound: 0.1
+  rewards_upper_bound: 0.9
+
+env:
```
Comment (F9rozen) [inline examples/embodiment/config/libero_spatial_grpo_evo1.yaml:82]: need align with evaluation :init_params:camera_heights: 448,camera_widths: 448

File: rlinf/models/embodiment/evo1/evo1_action_model.py [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+        images = [
+            [pixel_values[b, v] for v in range(num_views)] for b in range(batch_size)
+        ]
+        embodiment_id = torch.zeros(batch_size, dtype=torch.long, device=device)
+
+        autocast_enabled = device.type == "cuda"
+        with torch.autocast(
+            device_type="cuda", dtype=torch.bfloat16, enabled=autocast_enabled
+        ):
+            fused = self.evo1.get_vl_embeddings(
```
Comment (F9rozen) [inline rlinf/models/embodiment/evo1/evo1_action_model.py:460]: maybe put into "forward_inputs"

File: docs/source-en/rst_source/examples/embodied/evo1.rst [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    actor:
+      model:
+        model_type: "evo1"
+        model_path: "/path/to/model/Evo1_LIBERO"
+        num_action_chunks: 14        # steps executed per inference
+        evo1:
+          arm_key: "libero_robot"    # must match norm_stats.json
+          dataset_key: "libero_robot"
+        rl_head_config:
+          noise_method: "flow_sde"
```
Comment (F9rozen) [inline docs/source-en/rst_source/examples/embodied/evo1.rst]: never used

File: examples/embodiment/config/libero_spatial_grpo_evo1.yaml [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+  group_name: "EnvGroup"
+
+  train:
+    rollout_epoch: 8
+    total_num_envs: 64           # 64 // 8 GPUs = 8 env/worker -> group_size=8 valid
+    group_size: ${algorithm.group_size}
+    max_episode_steps: 252  # multiple of num_action_chunks=14 (14*18)
+    max_steps_per_rollout_epoch: 252  # divisible by num_action_chunks=14
```
Comment (F9rozen) [inline examples/embodiment/config/libero_spatial_grpo_evo1.yaml]: diff with evaluation

File: examples/embodiment/config/libero_spatial_grpo_evo1.yaml (hunk 1 above)
Comment (DorayakiLin) [reply examples/embodiment/config/libero_spatial_grpo_evo1.yaml:148]: Done — added use_orig_params: True to actor.fsdp_config. Thanks for catching the mismatch with the note in model/evo1.yaml.

File: examples/embodiment/config/libero_spatial_grpo_evo1.yaml (hunk 2 above)
Comment (DorayakiLin) [reply examples/embodiment/config/libero_spatial_grpo_evo1.yaml]: Done — replaced all machine-specific paths with /path/to/... placeholders across the grpo / sft / model/ eval configs (model_path → /path/to/Evo1_LIBERO, repo_path → /path/to/Evo-1), matching the other model configs' convention.

File: rlinf/models/embodiment/evo1/evo1_action_model.py (hunk 3 above)
Comment (DorayakiLin) [reply rlinf/models/embodiment/evo1/evo1_action_model.py]: Fixed — kept a single definition at the top of the class.

File: examples/embodiment/config/libero_spatial_grpo_evo1.yaml (hunk 4 above)
Comment (DorayakiLin) [reply examples/embodiment/config/libero_spatial_grpo_evo1.yaml:82]: Done — added init_params: { camera_heights: 448, camera_widths: 448 } to both env.train and env.eval to match the eval config. Evo-1 trains/evaluates at 448×448, so the RL rollout env now renders at the same resolution.

File: rlinf/models/embodiment/evo1/evo1_action_model.py (hunk 5 above)
Comment (DorayakiLin) [reply rlinf/models/embodiment/evo1/evo1_action_model.py:460]: Done — embodiment_id is now packed into forward_inputs (_pack_forward_inputs) and read back in default_forward, so rollout and replay share the same tensor instead of recomputing it inline.

File: docs/source-en/rst_source/examples/embodied/evo1.rst (hunk 6 above)
Comment (DorayakiLin) [reply docs/source-en/rst_source/examples/embodied/evo1.rst]: Fixed — removed noise_method from the Evo-1 config and docs.

File: examples/embodiment/config/libero_spatial_grpo_evo1.yaml (hunk 7 above)
Comment (DorayakiLin) [reply examples/embodiment/config/libero_spatial_grpo_evo1.yaml]: Aligned — set max_episode_steps / max_steps_per_rollout_epoch to 350 (14×25) to match the evaluation 
budget. total_num_envs intentionally differs (64 train throughput vs 120 eval coverage, as in the other configs)

**Review by F9rozen — APPROVED** [review]

**Review by thereAreDemonsNearby — APPROVED** [review]

File: requirements/install.sh [hunk 8]
```diff
@@ -1452,6 +1452,40 @@ install_starvla_model() {
...
+            install_${ENV_NAME}_env
+            ;;
+        *)
+            echo "Environment '$ENV_NAME' is not supported for Evo-1 model." >&2
+            exit 1
+            ;;
+    esac
+
+    local evo1_path
+    evo1_path=$(clone_or_reuse_repo EVO1_PATH "$VENV_DIR/Evo-1" https://github.com/MINT-SJTU/Evo-1.git -b "${EVO1_GIT_REF:-evo1-flash}" --depth 1)
```
Comment (Lin-xs) [inline requirements/install.sh]: Should the GitHub repository link passed to `clone_or_reuse_repo` be prefixed with `GITHUB_PREFIX`? I notice most commands in `install.sh` omit this prefix, whereas `install_lingbot_vla_model` includes it. Could you clarify the best practice and double-check this logic? @andylin-hao

File: rlinf/models/embodiment/evo1/evo1_action_model.py [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+            getattr(evo1_model, "action_head", None), "num_inference_timesteps", None
+        )
+        self.rl_denoising_steps = int(
+            rl_denoising_steps
+            or rlh.cfg_get(rl_head_config, "denoising_steps", None)
+            or head_steps
+            or 32
+        )
+
+    # ------------------------------------------------------------------ #
```
Comment (Lin-xs) [inline rlinf/models/embodiment/evo1/evo1_action_model.py]: Please delete these comments.

File: rlinf/models/embodiment/evo1/evo1_rl_head.py [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+
+from __future__ import annotations
+
+import random
+from typing import Any, Literal, Optional
+
+import torch
+
+
+# ---------------------------------------------------------------------------- #
```
Comment (Lin-xs) [inline rlinf/models/embodiment/evo1/evo1_rl_head.py]: Same issue.

File: requirements/install.sh (hunk 8 above)
Comment (andylin-hao) [reply requirements/install.sh]: No need to add the prefix. The git clone is already configured to go through ghfast if --use-mirror is specified

File: rlinf/models/embodiment/evo1/evo1_action_model.py (hunk 9 above)
Comment (DorayakiLin) [reply rlinf/models/embodiment/evo1/evo1_action_model.py]: Done — removed the `# ---- #` decorative section-separator comment blocks throughout evo1_action_model.py (5 blocks). Pushed in 4f807079.

File: rlinf/models/embodiment/evo1/evo1_rl_head.py (hunk 10 above)
Comment (DorayakiLin) [reply rlinf/models/embodiment/evo1/evo1_rl_head.py]: Done — same cleanup in evo1_rl_head.py, removed all 7 decorative separator comment blocks. 4f807079.

File: requirements/install.sh (hunk 8 above)
Comment (DorayakiLin) [reply requirements/install.sh]: Thanks for the clarification! The Evo-1 clone already passes the plain GitHub URL to clone_or_reuse_repo without GITHUB_PREFIX, so it routes through ghfast automatically when --use-mirror is set. Leaving it as-is — no change needed.

**Review by andylin-hao — APPROVED** [review]

**Review by guozhen1997 — APPROVED** [review]
